Expose prometheus compatible metrics on /prometheus

klette opened the pull request:
We run Zipkin in several kubernetes clusters and thus we need a sane way of collecting the metrics.
As Prometheus has all the built-in support for dealing with the moving parts of Kubernetes, it seems
like a rational choice.

This PR exposes Spring Actuator metrics on `/prometheus`
using the 0.0.4 version of their text format specified
here https://prometheus.io/docs/instrumenting/exposition_formats/.

Not crazy about the path, but adding new controllers into /metrics/\* seems like more complexity than needed.

Commit:
Expose prometheus compatible metrics on /prometheus

Exposes Spring Actuator metrics on `/prometheus`
using the 0.0.4 version of their text format specified
here https://prometheus.io/docs/instrumenting/exposition_formats/.

eirslett commented:
@negatioN interesting for FINN?

codefromthecrypt commented:
cool. I didn't expect this to be possible without adding a dependency (though we ought to have a test dependency to ensure the results are parseable by prometheus).

Metrics export is something that is quite deployment-specific.. would you mind creating an autoconfigure module for this instead? you can then make that module an optional dependency of zipkin-server. This allows it to get into the all-jar (for docker etc), but not pulled in automatically for those running custom builds.

You could do still do everything in one class, just it reduces the scope a bit. That module should include a README saying what version of prometheus format it works with, and ideally a test parser.

I think this is fair to add to core as modularity in docker is still not worked out, this is only a small dependency, and assuming @kristofa also wants this, we have 2 core committers who work at different companies interested in prometheus (and might be able to help support bug-fixes).

Sound reasonable?

kristofa commented:
Prometheus support would be great.
Definitely interested.

Op zondag 26 juni 2016 heeft Adrian Cole notifications@github.com het
volgende geschreven:

> cool. I didn't expect this to be possible without adding a dependency
> (though we ought to have a test dependency to ensure the results are
> parseable by prometheus).
> 
> Metrics export is something that is quite deployment-specific.. would you
> mind creating an autoconfigure module for this instead? you can then make
> that module an optional dependency of zipkin-server. This allows it to get
> into the all-jar (for docker etc), but not pulled in automatically for
> those running custom builds.
> 
> You could do still do everything in one class, just it reduces the scope a
> bit. That module should include a README saying what version of prometheus
> format it works with, and ideally a test parser.
> 
> I think this is fair to add to core as modularity in docker is still not
> worked out, this is only a small dependency, and assuming @kristofa
> https://github.com/kristofa also wants this, we have 2 core committers
> who work at different companies interested in prometheus (and might be able
> to help support bug-fixes).
> 
> Sound reasonable?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/openzipkin/zipkin/pull/1144#issuecomment-228582560,
> or mute the thread
> https://github.com/notifications/unsubscribe/ACHltIkxEg-ZV5_F2v9tvzp48Gt8FRNFks5qPfVqgaJpZM4I9y5e
> .

klette commented:
Alright-y,

moved the code into a auto configure module, and added some docs. Couldn't find any Java based parser, and building and running go code as a part of the test suite seemed a bit overkill :)

codefromthecrypt commented on zipkin-autoconfigure/metrics-prometheus/src/main/java/zipkin/autoconfigure/metrics/PrometheusMetricsAutoConfiguration.java:
```diff
@@ -0,0 +1,102 @@
+    private Collection<PublicMetrics> publicMetrics;
+
+    @Autowired
+    public PrometheusMetricsAutoConfiguration(final Collection<PublicMetrics> publicMetrics) {
```
nit: if you're looking for final, I'd move it to the field

klette replied:
Fixed :)

codefromthecrypt commented on zipkin-autoconfigure/metrics-prometheus/src/main/java/zipkin/autoconfigure/metrics/PrometheusMetricsAutoConfiguration.java:
```diff
@@ -0,0 +1,102 @@
+    public ResponseEntity<String> prometheusMetrics() {
+        StringBuilder sb = new StringBuilder();
+
+        for (PublicMetrics publicMetrics : this.publicMetrics) {
```
interesting.. so this isn't just zipkin, this is all spring metrics!

codefromthecrypt commented on zipkin-autoconfigure/metrics-prometheus/src/test/java/zipkin/autoconfigure/metrics/PrometheusMetricsAutoConfigurationTest.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Test
+    public void correctHttpResponse() throws Exception {
+        final PublicMetrics publicMetrics = () -> Collections.singleton(new Metric<Number>("mem.free", 1024));
```
no big deal, but we don't put final on variables even if it is explicit and there are good reasons why some folks make conventions like that.

leaving them out keeps line shorter and focuses more attention on the test code (at risk of someone doing the shell-game with references)

klette replied:
Ah, cool. Just my brain that does that automatically after x years with that as a coding standard :)

codefromthecrypt commented on zipkin-autoconfigure/metrics-prometheus/README.md:
```diff
@@ -0,0 +1,18 @@
+# Prometheus Metrics
```
should probably make a link from zipkin-server's README to this, under the metrics section. Saying something like: Zipkin also exports metrics in prometheus format (link)

klette replied:
Done!

codefromthecrypt commented:
No required changes on my end, except linking to the server's README (to make it more discoverable)

nice work!

Commit:
Add small note and link to prometheus metrics from zipkin-server README

Commit:
Make `final` usage in metrics-prometheus consistent with the zipkin-project

codefromthecrypt commented:
nice work. squashed on entry

codefromthecrypt commented:
hi all. I just noticed that prometheus' java client now ships with spring boot support. any thoughts on whether we should switch? https://github.com/prometheus/client_java/tree/master/simpleclient_spring_boot/src/main/java/io/prometheus/client/spring/boot

cc @shakuzen
